src/azure-cli/azure/cli/command_modules/resource/custom.py

```diff
@@ -1055,6 +1055,168 @@ def _get_deployment_management_client(cli_ctx, aux_subscriptions=None, aux_tenan
+            name)
+        build_confirmation_string += "The following actions will be applied to any resources that are no longer managed by the deployment stack after the template is applied:\n"
+        # first case we have only detach
+        if delete_resources_enum == detach_model and delete_resource_groups_enum == detach_model:
+            build_confirmation_string += "\nDetach: resources and resource groups\n"
+        # second case we only have delete
+        elif delete_resources_enum == delete_model and delete_resource_groups_enum == delete_model:
+            build_confirmation_string += "\nDeleting: resources and resource groups\n"
+        else:
+            if delete_resources_enum == detach_model:
+                build_confirmation_string += "\nDetach: resources\n"
+                build_confirmation_string += "\nDeleting: resource groups\n"
+            else:
+                build_confirmation_string += "\nDetach: resource groups\n"
+                build_confirmation_string += "\nDeleting: resources\n"
+        confirmation = prompt_y_n(build_confirmation_string + "\n")
+        if not confirmation:
+            return None
+
+    return build_confirmation_string
+
+
+def _prepare_stacks_templates_and_parameters(cmd, rcf, deployment_stack_model, template_file, template_spec, template_uri, parameters, query_string):
+    t_spec, t_uri = None, None
+    template_obj = None
+
+    deployment_stacks_template_link = rcf.deployment_stacks.models.DeploymentStacksTemplateLink()
+
+    if template_file:
+        pass
+    elif template_spec:
+        t_spec = template_spec
+    elif template_uri:
+        t_uri = template_uri
+    else:
+        raise InvalidArgumentValueError(
+            "Please enter one of the following: template file, template spec, or template url")
+
+    if t_spec:
+        deployment_stacks_template_link.id = t_spec
+        deployment_stack_model.template_link = deployment_stacks_template_link
+        api_version = get_api_version(cmd.cli_ctx, ResourceType.MGMT_RESOURCE_TEMPLATESPECS)
+        template_obj = show_resource(cmd=cmd, resource_ids=[template_spec],
+                                     api_version=api_version).properties['mainTemplate']
+    elif t_uri:
+        if query_string:
+            deployment_stacks_template_link = rcf.deployment_stacks.models.DeploymentStacksTemplateLink(
+                uri=t_uri, query_string=query_string)
+            t_uri = _prepare_template_uri_with_query_string(
+                template_uri=t_uri, input_query_string=query_string)
+        else:
+            deployment_stacks_template_link = rcf.deployment_stacks.models.DeploymentStacksTemplateLink(uri=t_uri)
+        deployment_stack_model.template_link = deployment_stacks_template_link
+        template_obj = _remove_comments_from_json(_urlretrieve(t_uri).decode('utf-8'), file_path=t_uri)
+    else:
+        if _is_bicepparam_file_provided(parameters):
+            ensure_bicep_installation(cmd.cli_ctx)
```

Comment:
> E   vcr.errors.CannotOverwriteExistingCassetteException: Can't overwrite existing cassette ('/mnt/vss/_work/1/s/src/azure-cli/azure/cli/command_modules/resource/tests/latest/recordings/test_create_deployment_stack_resource_group.yaml') in your current record mode ('once').
>    No match for the request (<Request (GET) https://downloads.bicep.azure.com/v0.18.4/bicep-linux-x64>) was found.
    Found 1 similar requests with 1 different matcher(s) :
>    1 - (<Request (GET) https://downloads.bicep.azure.com/releases/latest>).
    Matchers succeeded : ['method', 'scheme', 'host', 'port', '_custom_request_query_matcher']
    Matchers failed :
    path - assertion failure :
    /v0.18.4/bicep-linux-x64 != /releases/latest


In the CI pipeline, since the `ensure_bicep_installation` method uses the `get_bicep_latest_release_tag` method to obtain the release tag for Bicep CLI, the current release tag from `https://aka.ms/BicepLatestRelease` is `v0.18.4`.

Currently, the request from `urlopen(bicep_download_url)` for downloading Bicep CLI is not recorded in the recording file, which has caused the CI issue mentioned above.

https://github.com/Azure/azure-cli/blob/5157cf9be2336377c0383cfd4ca193c52ed3c053/src/azure-cli/azure/cli/command_modules/resource/_bicep.py#L128-L137

Additionally, I see that currently all tests related to Bicep are based on `LiveScenarioTest`, as in order to avoid recording binary data of bicep CLI, we should keep consistency in the bicep related tests, such as using `@live_only()` or `LiveScenarioTest`

https://github.com/Azure/azure-cli/blob/5157cf9be2336377c0383cfd4ca193c52ed3c053/src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py#L4046-L4047

Reply:
@shenglol By the way, may I ask do you have any suggestion for writing the Bicep related tests?

Reply:
Yeah, live tests make sense as we don't want to spam the recording files with Bicep CLI binary data.